Marksman/Toc.fs

```diff
@@ -0,0 +1,94 @@
+type EntryLevel = int
+
+type Entry = { Level: EntryLevel; Title: Title; Link: Slug }
+
+module Entry =
+    let renderLink entry =
+        let offset = String.replicate (entry.Level - 1) " "
+        let slug = entry.Link |> Slug.toString
+        $"{offset}- [{entry.Title}](#{slug})"
+
+    let fromHeading (heading: Heading) : Entry =
+        let slug = Heading.slug heading
+        { Level = heading.level; Link = slug; Title = heading.title.text }
+
+type TableOfContents = { Entries: Entry[] }
+
+module TableOfContents =
+    let logger = LogProvider.getLoggerByName "TocAgent"
+
+    let mk (index: Index) : TableOfContents option =
+        let headings = index.headings |> Array.map (fun x -> x.data)
+
+        if index.headings.Length.Equals 0 then
+            None
+        else
+            Some { Entries = Array.map Entry.fromHeading headings }
+
+    let render (toc: TableOfContents) =
+        Marker
+        + "\n"
+        + (Array.map Entry.renderLink toc.Entries |> String.concat "\n")
+        + "\n"
+        + "\n"
+        + "\n"
+
+    type State =
+        | BeforeMarker
+        | Collecting of Range * int
+
+
+    let detect (text: Text) : Range option =
+        let lines = text.lineMap
+        let maxIndex = lines.NumLines
+
+        let rec go i st =
+            if i.Equals(maxIndex) then
+                st
+            else
+                let lineRange = text.LineContentRange(i)
+                let lineContent = text.LineContent(i)
+                let lineIsEmpty = lineContent.Trim().Length.Equals(0)
+                let lineIsMarker = lineContent.Equals(Marker)
+                let expandToThisLine (range: Range) = { range with End = lineRange.End }
+
+                match st with
+                // if we found the marker, start collecting text from here
+                | BeforeMarker when lineIsMarker -> go (i + 1) (Collecting(lineRange, 0))
+                // if we are in collecting mode, just expand the selection
+                | Collecting (range, numEmptyLines) ->
+                    let toThisLine = expandToThisLine range
+
+                    if lineIsEmpty then
+                        if numEmptyLines.Equals(0) then
+                            go (i + 1) (Collecting(toThisLine, 1))
+                        else
+                            // extend to next line
+                            let extraLine =
+                                let last = toThisLine.End
+                                { toThisLine with End = { Line = last.Line + 1; Character = 0 } }
+
+                            Collecting(extraLine, 2)
+                    else
+                        go (i + 1) (Collecting(expandToThisLine range, 0))
+                | BeforeMarker -> go (i + 1) BeforeMarker
+
+        match (go 0 BeforeMarker) with
+        // marker was never found
+        | Collecting (range, 2) -> Some range
+        | other ->
+            logger.error (Log.setMessage $"Got {other} instead")
```

Comment:
maybe warn?

also you could do `Log.setMessage "[what went wrong]" >> Log.addContext "state" other` -- structured logging FTW